TST: 27 test failures with numpy 2.1.dev

Example log: https://github.com/spacetelescope/spherical_geometry/actions/runs/8595276994/job/23549707417

Example traceback (one of many):

```
>       A = polygon.SphericalPolygon(a)

spherical_geometry/tests/test_union.py:329: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
spherical_geometry/polygon.py:772: in __init__
    polygons.extend(g.disjoint_polygons())
spherical_geometry/graph.py:458: in disjoint_polygons
    changed = self._find_all_intersections() or changed
spherical_geometry/graph.py:643: in _find_all_intersections
    changed = self._find_point_to_arc_intersections() or changed
spherical_geometry/graph.py:542: in _find_point_to_arc_intersections
    intersects = gca.intersects_point(
spherical_geometry/great_circle_arc.py:286: in intersects_point
    total_length = length(A, B)
spherical_geometry/great_circle_arc.py:227: in length
    dot = inner1d(A, B)
spherical_geometry/great_circle_arc.py:31: in _inner1d_np
    return np.multiply(x, y).sum(axis=1)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

a = array([0.09374197, 0.00091684, 0.90534119]), axis = 1, dtype = None
out = None, keepdims = False, initial = <no value>, where = True

    def _sum(a, axis=None, dtype=None, out=None, keepdims=False,
             initial=_NoValue, where=True):
>       return umr_sum(a, axis, dtype, out, keepdims, initial, where)
E       numpy.exceptions.AxisError: axis 1 is out of bounds for array of dimension 1
```

--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -7,7 +7,13 @@ Release Notes
 1.3.2 (unreleased)
 ==================
 
-- Fixed a bug in the Python implementation of ``inner1d``. [#266]
+- Fixed a bug in the Python implementation of ``inner1d``. [#265]
+
+- Removed the ``degree`` argument from the ``length`` and ``angle`` functions
+  in the ``spherical_geometry.great_circle_arc`` module as it was never working
+  correctly with the ``math_util`` module and add unnecessary
+  complication. [#265]
+
 
 1.3.1 (5-November-2023)
 =======================

--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -4,6 +4,11 @@
 Release Notes
 =============
 
+1.3.2 (unreleased)
+==================
+
+- Fixed a bug in the Python implementation of ``inner1d``. [#266]
+
 1.3.1 (5-November-2023)
 =======================
 

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -53,8 +53,8 @@ documentation = "http://spherical-geometry.readthedocs.io/"
 requires = [
     "setuptools>=61.2",
     "setuptools_scm[toml]>=3.6",
-    "numpy>=1.25",
-#    "numpy>=2.0.0dev0",
+    #"numpy>=1.25",
+    "numpy>=2.0.0dev0",
 ]
 build-backend = "setuptools.build_meta"
 

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -53,8 +53,8 @@ documentation = "http://spherical-geometry.readthedocs.io/"
 requires = [
     "setuptools>=61.2",
     "setuptools_scm[toml]>=3.6",
-    #"numpy>=1.25",
-    "numpy>=2.0.0dev0",
+    "numpy>=1.25",
+    #"numpy>=2.0.0dev0",
 ]
 build-backend = "setuptools.build_meta"
 

--- a/spherical_geometry/great_circle_arc.py
+++ b/spherical_geometry/great_circle_arc.py
@@ -23,12 +23,12 @@ try:
 except ImportError:
     HAS_C_UFUNCS = False
 
-__all__ = ['angle', 'intersection', 'intersects', 'intersects_point',
-           'length', 'midpoint', 'interpolate']
+__all__ = ['angle', 'interpolate', 'intersection', 'intersects',
+           'intersects_point', 'length', 'midpoint']
 
 
 def _inner1d_np(x, y):
-    return np.multiply(x, y).sum(axis=1)
+    return np.multiply(x, y).sum(axis=-1)
 
 
 if HAS_C_UFUNCS:
@@ -213,6 +213,8 @@ def length(A, B, degrees=True):
     if HAS_C_UFUNCS:
         result = math_util.length(A, B)
     else:
+        raise AssertionError("C version should have been used")
+        approx1 = 1 + 3 * np.finfo(float).eps
         A = np.asanyarray(A)
         B = np.asanyarray(B)
 
@@ -221,11 +223,20 @@ def length(A, B, degrees=True):
         B2 = B ** 2.0
         Bl = np.sqrt(np.sum(B2, axis=-1))
 
-        A = A / two_d(Al)
-        B = B / two_d(Bl)
+        try:
+            with np.errstate(invalid='raise'):
+                A = A / two_d(Al)
+                B = B / two_d(Bl)
+        except FloatingPointError:
+            raise ValueError("Out of domain for acos")
 
         dot = inner1d(A, B)
-        dot = np.clip(dot, -1.0, 1.0)
+
+        for d in np.atleast_1d(dot):
+            if np.isnan(d) or abs(d) > approx1:
+                raise ValueError("Out of domain for acos")
+
+        dot = np.clip(dot, -1.0, 1.0)  # needed due to accuracy loss
         with np.errstate(invalid='ignore'):
             result = np.arccos(dot)
 

--- a/spherical_geometry/great_circle_arc.py
+++ b/spherical_geometry/great_circle_arc.py
@@ -19,8 +19,7 @@ from spherical_geometry.vector import two_d
 # the python versions are a fallback if the C cannot be used
 try:
     from spherical_geometry import math_util
-    math_util = None
-    HAS_C_UFUNCS = False
+    HAS_C_UFUNCS = True
 except ImportError:
     HAS_C_UFUNCS = False
 

--- a/spherical_geometry/great_circle_arc.py
+++ b/spherical_geometry/great_circle_arc.py
@@ -196,7 +196,7 @@ def length(A, B):
     Returns
     -------
     length : scalar or array of scalars
-        The angular length of the great circle arc.
+        The angular length of the great circle arc in radians.
 
     Notes
     -----

--- a/spherical_geometry/great_circle_arc.py
+++ b/spherical_geometry/great_circle_arc.py
@@ -19,7 +19,8 @@ from spherical_geometry.vector import two_d
 # the python versions are a fallback if the C cannot be used
 try:
     from spherical_geometry import math_util
-    HAS_C_UFUNCS = True
+    math_util = None
+    HAS_C_UFUNCS = False
 except ImportError:
     HAS_C_UFUNCS = False
 
@@ -183,7 +184,7 @@ def intersection(A, B, C, D):
     return np.where(equals, np.nan, cross)
 
 
-def length(A, B, degrees=True):
+def length(A, B):
     r"""
     Returns the angular distance between two points (in vector space)
     on the unit sphere.
@@ -192,10 +193,6 @@ def length(A, B, degrees=True):
     ----------
     A, B : (*x*, *y*, *z*) triples or Nx3 arrays of triples
        The endpoints of the great circle arc, in vector space.
-
-    degrees : bool, optional
-        If `True` (default) the result is returned in decimal degrees,
-        otherwise radians.
 
     Returns
     -------
@@ -213,7 +210,6 @@ def length(A, B, degrees=True):
     if HAS_C_UFUNCS:
         result = math_util.length(A, B)
     else:
-        raise AssertionError("C version should have been used")
         approx1 = 1 + 3 * np.finfo(float).eps
         A = np.asanyarray(A)
         B = np.asanyarray(B)
@@ -240,10 +236,7 @@ def length(A, B, degrees=True):
         with np.errstate(invalid='ignore'):
             result = np.arccos(dot)
 
-    if degrees:
-        return np.rad2deg(result)
-    else:
-        return result
+    return result
 
 
 def intersects(A, B, C, D):
@@ -300,10 +293,10 @@ def intersects_point(A, B, C):
 
     length_diff = np.abs((left_length + right_length) - total_length)
 
-    return length_diff < 1e-10
+    return length_diff < 3e-11
 
 
-def angle(A, B, C, degrees=True):
+def angle(A, B, C):
     """
     Returns the angle at *B* between *AB* and *BC*.
 
@@ -311,10 +304,6 @@ def angle(A, B, C, degrees=True):
     ----------
     A, B, C : (*x*, *y*, *z*) triples or Nx3 arrays of triples
         Points on sphere.
-
-    degrees : bool, optional
-        If `True` (default) the result is returned in decimal degrees,
-        otherwise radians.
 
     Returns
     -------
@@ -339,14 +328,20 @@ def angle(A, B, C, degrees=True):
         ABX = _cross_and_normalize(A, B)
         BCX = _cross_and_normalize(C, B)
         X = _cross_and_normalize(ABX, BCX)
+        m = np.logical_or(
+                np.linalg.norm(ABX, axis=-1) == 0.0,
+                np.linalg.norm(BCX, axis=-1) == 0.0
+            )
+
         diff = inner1d(B, X)
         inner = inner1d(ABX, BCX)
         with np.errstate(invalid='ignore'):
+            inner = np.clip(inner, -1.0, 1.0)  # needed due to accuracy loss
             angle = np.arccos(inner)
+
         angle = np.where(diff < 0.0, (2.0 * np.pi) - angle, angle)
 
-    if degrees:
-        angle = np.rad2deg(angle)
+        angle[m] = np.nan
 
     return angle
 
@@ -405,7 +400,7 @@ def interpolate(A, B, steps=50):
     steps = int(max(steps, 2))
     t = np.linspace(0.0, 1.0, steps, endpoint=True).reshape((steps, 1))
 
-    omega = length(A, B, degrees=False)
+    omega = length(A, B)
     if omega == 0.0:
         offsets = t
     else:

--- a/spherical_geometry/polygon.py
+++ b/spherical_geometry/polygon.py
@@ -546,8 +546,7 @@ class SingleSphericalPolygon(object):
             return np.array(0.0)
 
         points = np.vstack((self._points, self._points[1]))
-        angles = great_circle_arc.angle(points[:-2], points[1:-1],
-                                        points[2:], degrees=False)
+        angles = great_circle_arc.angle(points[:-2], points[1:-1], points[2:])
 
         return np.sum(angles) - (len(angles) - 2) * np.pi
 
@@ -704,7 +703,7 @@ class SingleSphericalPolygon(object):
 
         alpha = 1.0
         for A, B in zip(points[0:-1], points[1:]):
-            length = great_circle_arc.length(A, B, degrees=True)
+            length = np.rad2deg(great_circle_arc.length(A, B))
             if not np.isfinite(length):
                 length = 2
             interpolated = great_circle_arc.interpolate(A, B, length * 4)

--- a/spherical_geometry/tests/test_basic.py
+++ b/spherical_geometry/tests/test_basic.py
@@ -161,8 +161,6 @@ def test_interpolate():
         assert abs(length - first_length) < 1.0e-10
 
 
-@pytest.mark.xfail(
-    math_util is None, reason="math_util C-ext is missing, numpy gives different results")
 def test_overlap():
     def build_polygon(offset):
         points = []
@@ -380,26 +378,26 @@ def test_great_circle_arc_length():
     B = [-90, 0]
     A = vector.lonlat_to_vector(*A)
     B = vector.lonlat_to_vector(*B)
-    assert great_circle_arc.length(A, B) == 180.0
+    assert_almost_equal(great_circle_arc.length(A, B), np.pi)
 
     A = [135, 0]
     B = [-90, 0]
     A = vector.lonlat_to_vector(*A)
     B = vector.lonlat_to_vector(*B)
-    assert_almost_equal(great_circle_arc.length(A, B), 135.0)
+    assert_almost_equal(great_circle_arc.length(A, B), (3.0 / 4.0) * np.pi)
 
     A = [0, 0]
     B = [0, 90]
     A = vector.lonlat_to_vector(*A)
     B = vector.lonlat_to_vector(*B)
-    assert_almost_equal(great_circle_arc.length(A, B), 90.0)
+    assert_almost_equal(great_circle_arc.length(A, B), np.pi / 2.0)
 
 
 def test_great_circle_arc_angle():
     A = [1, 0, 0]
     B = [0, 1, 0]
     C = [0, 0, 1]
-    assert great_circle_arc.angle(A, B, C) == 270.0
+    assert great_circle_arc.angle(A, B, C) == (3.0 / 2.0) * np.pi
 
     # TODO: More angle tests
 
@@ -522,10 +520,9 @@ def test_convex_hull(repeat_pts):
         assert b == r, "Polygon boundary has correct points"
 
 
-@pytest.mark.skipif(math_util is None, reason="math_util C-ext is missing")
 def test_math_util_angle_domain():
     assert not np.isfinite(
-        math_util.angle([[0, 0, 0]], [[0, 0, 0]], [[0, 0, 0]])[0]
+        great_circle_arc.angle([[0, 0, 0]], [[0, 0, 0]], [[0, 0, 0]])[0]
     )
 
 
@@ -534,7 +531,6 @@ def test_math_util_length_domain():
         great_circle_arc.length([[np.nan, 0, 0]], [[0, 0, np.inf]])
 
 
-@pytest.mark.skipif(math_util is None, reason="math_util C-ext is missing")
 def test_math_util_angle_nearly_coplanar_vec():
     # test from issue #222 + extra values
     vectors = [
@@ -542,7 +538,7 @@ def test_math_util_angle_nearly_coplanar_vec():
         5 * [[1, 0.9999999, 1]],
         [[1, 0.5, 1], [1, 0.15, 1], [1, 0.001, 1], [1, 0.15, 1], [-1, 0.1, -1]]
     ]
-    angles = math_util.angle(*vectors)
+    angles = great_circle_arc.angle(*vectors)
 
     assert_allclose(angles[:-1], np.pi, rtol=0, atol=1e-16)
     assert_allclose(angles[-1], 0, rtol=0, atol=1e-32)

--- a/spherical_geometry/tests/test_basic.py
+++ b/spherical_geometry/tests/test_basic.py
@@ -162,14 +162,16 @@ def test_interpolate():
 
 
 def test_overlap():
+    y_eps = 1e-8
+
     def build_polygon(offset):
         points = []
         corners = [(0.0, 0.0), (0.0, 10.0), (10.0, 10.0), (10.0, 0.0)]
         for corner in corners:
-            point = np.asarray(corner)
-            point[0] += offset
-            points.append(np.asarray(vector.lonlat_to_vector(point[0],
-                                                             point[1])))
+            x, y = corner
+            points.append(
+                np.asarray(vector.lonlat_to_vector(x + offset, y + y_eps))
+            )
         poly = polygon.SphericalPolygon(points)
         return poly
 

--- a/spherical_geometry/tests/test_basic.py
+++ b/spherical_geometry/tests/test_basic.py
@@ -529,10 +529,9 @@ def test_math_util_angle_domain():
     )
 
 
-@pytest.mark.skipif(math_util is None, reason="math_util C-ext is missing")
 def test_math_util_length_domain():
     with pytest.raises(ValueError):
-        math_util.length([[np.nan, 0, 0]], [[0, 0, np.inf]])
+        great_circle_arc.length([[np.nan, 0, 0]], [[0, 0, np.inf]])
 
 
 @pytest.mark.skipif(math_util is None, reason="math_util C-ext is missing")

--- a/spherical_geometry/tests/test_union.py
+++ b/spherical_geometry/tests/test_union.py
@@ -18,6 +18,7 @@ from spherical_geometry.tests.helpers import ROOT_DIR, resolve_imagename
 
 try:
     from spherical_geometry import math_util
+    math_util = None
 except ImportError:
     math_util = None
 
@@ -332,9 +333,11 @@ def test_edge_crossings():
 
 
 @pytest.mark.skipif(
-    math_util is None, reason="math_util C-ext is missing, double accuracy leads to crash")
+    math_util is None,
+    reason="math_util C-ext is missing, double accuracy leads to crash"
+)
 def test_almost_identical_polygons_multi_union():
-    filename = resolve_imagename(ROOT_DIR,'almost_same_polygons.npz')
+    filename = resolve_imagename(ROOT_DIR, 'almost_same_polygons.npz')
     polygon_data = np.load(filename)
 
     polygons = []

--- a/spherical_geometry/tests/test_union.py
+++ b/spherical_geometry/tests/test_union.py
@@ -18,7 +18,6 @@ from spherical_geometry.tests.helpers import ROOT_DIR, resolve_imagename
 
 try:
     from spherical_geometry import math_util
-    math_util = None
 except ImportError:
     math_util = None
 

--- a/spherical_geometry/tests/test_union.py
+++ b/spherical_geometry/tests/test_union.py
@@ -331,8 +331,8 @@ def test_edge_crossings():
     _ = A.union(B)
 
 
-@pytest.mark.xfail(
-    math_util is None, reason="math_util C-ext is missing, numpy gives different results")
+@pytest.mark.skipif(
+    math_util is None, reason="math_util C-ext is missing, double accuracy leads to crash")
 def test_almost_identical_polygons_multi_union():
     filename = resolve_imagename(ROOT_DIR,'almost_same_polygons.npz')
     polygon_data = np.load(filename)

--- a/src/math_util.c
+++ b/src/math_util.c
@@ -71,6 +71,8 @@ typedef struct {
     double x[4];
 } qd;
 
+#define ISNAN_QD(q) ((q.x[0]) != (q.x[0]))
+
 double QD_ONE[4] = {1.0, 0.0, 0.0, 0.0};
 
 static NPY_INLINE void
@@ -234,7 +236,15 @@ sign(const double A) {
 
 static NPY_INLINE int
 equals_qd(const qd *A, const qd *B) {
-    return memcmp(A, B, sizeof(qd) * 3) == 0;
+    if (memcmp(A, B, sizeof(qd) * 3)) {
+        return 0;
+    }
+    for (int k = 0; k < 3; ++k) {
+        if (ISNAN_QD(A[k])) {
+            return 0;
+        }
+    }
+    return 1;
 }
 
 static NPY_INLINE int
@@ -252,7 +262,7 @@ length_qd(const qd *A, const qd *B, qd *l) {
 
     dot_qd(A, B, &s);
 
-    if (s.x[0] != s.x[0] ||
+    if (ISNAN_QD(s) ||
         s.x[0] < -1.0 ||
         s.x[0] > 1.0) {
         PyErr_SetString(PyExc_ValueError, "Out of domain for acos");

